tools/toolchains/mbed_toolchain.py

```diff
@@ -397,14 +397,27 @@ def get_arch_file(self, objects):
 
     # THIS METHOD IS BEING CALLED BY THE MBED ONLINE BUILD SYSTEM
     # ANY CHANGE OF PARAMETERS OR RETURN VALUES WILL BREAK COMPATIBILITY
-    def compile_sources(self, resources, inc_dirs=None):
+    def compile_sources(self, resources, inc_dirs=None, exclude_paths=None):
```

Comment:
This addition of the `exclude_paths` parameter will mean that it will be included the statement:

```
    # THIS METHOD IS BEING CALLED BY THE MBED ONLINE BUILD SYSTEM
    # ANY CHANGE OF PARAMETERS OR RETURN VALUES WILL BREAK COMPATIBILITY
```

It would be possible to remove this parameter when legacy build support is removed but I'm sure by then we would have forgotten. My suggestion would be to create a private function which has this parameter but leave this interface as it is, e.g. 

```
def compile_legacy_sources(self, resources, inc_dirs=None, exclude_paths=None)
    return _compile_sources(resources, inc_dirs=inc_dirs, exclude_paths=exclude_paths)

def compile_sources(self, resources, inc_dirs=None)
    return _compile_sources(resources, inc_dirs=inc_dirs)

def _compile_sources(self, resources, inc_dirs=None, exclude_paths=None)
    ...
```
I would also ask whether there is a different between `dirs` and `paths` the two different terminologies in the parameter list might be confusing.